headerforwarder/forwarder.go

```diff
@@ -0,0 +1,123 @@
+package headerforwarder
+
+import (
+	"fmt"
+	"net/http"
+)
+
+// HeaderExtractingTransport is a utility to help a rosetta server forward headers to and from
+// native node requests. It implements several interfaces to achieve that:
+//   - http.RoundTripper: this can be used to create an http Client that will automatically save headers
+//     if necessary
+//
+// the headers can be requested later.
+//
+// TODO: this needs to expire entries after a certain amount of time
+type HeaderForwarder struct {
+	requestHeaders     map[string]http.Header
+	interestingHeaders []string
+	actualTransport    http.RoundTripper
+}
+
+func NewHeaderForwarder(interestingHeaders []string, transport http.RoundTripper) *HeaderForwarder {
+	return &HeaderForwarder{
+		requestHeaders:     make(map[string]http.Header),
+		interestingHeaders: interestingHeaders,
+		actualTransport:    transport,
+	}
+}
+
+// RoundTrip implements http.RoundTripper and will be used to construct an http Client which
+// saves the native node response headers if necessary.
+func (hf *HeaderForwarder) RoundTrip(req *http.Request) (*http.Response, error) {
+	resp, err := hf.actualTransport.RoundTrip(req)
+
+	if err == nil && hf.shouldRememberHeaders(req, resp) {
+		hf.rememberHeaders(req, resp)
+	} else {
+		fmt.Println("not remembering headers", resp.Header)
+	}
+
+	return resp, err
+}
+
+// shouldRememberHeaders is called to determine if response headers should be remembered for a given request.
+// Response headers will only be remembered if the request does not contain all of the interesting
```

Comment:
What if a sticky session cookie is provided in the request but it expires and a new sticky session cookie comes in the response?

Reply:
technically since the `Cookie` and `Set-Cookie` header are different, headers for requests with cookies will always be forwarded.

If that's eventually not what we want we'd have to write custom logic to associate `Cookie` and `Set-Cookie` together for request/response (and storing/accessing) respectively